src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -397,23 +397,21 @@ public void ExecuteSuppressionAction(DiagnosticSuppressor suppressor, ImmutableA
         {
             Debug.Assert(compilationEvent is CompilationStartedEvent || compilationEvent is CompilationCompletedEvent);
 
-            AnalyzerStateData? analyzerState = null;
-
-            try
+            if (TryStartProcessingEvent(compilationEvent, analyzer, analysisScope, analysisState, out var analyzerState))
             {
-                if (TryStartProcessingEvent(compilationEvent, analyzer, analysisScope, analysisState, out analyzerState))
+                try
                 {
                     ExecuteCompilationActionsCore(compilationActions, analyzer, analyzerState);
-                    analysisState?.MarkEventComplete(compilationEvent, analyzer);
-                    return true;
+                }
+                finally
+                {
+                    analyzerState?.ResetToReadyState();
```

Comment:
>analyzerState?.ResetToReadyState(); [](http://example.com/codeflow?start=20&length=35)

Is the following going to decrease/eliminate the chances of us failing to complete the event because another thread "hijacked" the state after this line?

We could introduce another state, called "AttemptingToComplete". Before executing ```TryMarkEventComplete```, we don't reset the state to "ReadyToProcess", instead we change it to ```AttemptingToComplete```. And only, if ```TryMarkEventComplete``` failed, we reset to "ReadyToProcess".

I am not sure if this is going to introduce too much complexity, or, perhaps, a race for completion between different threads.

Reply:
> Is the following going to decrease/eliminate the chances of us failing to complete the event because another thread "hijacked" the state after this line?

Yes, this definitely can happen.

> I am not sure if this is going to introduce too much complexity, or, perhaps, a race for completion between different threads.

That is my primary concern. Let me open a separate tracking issue to try the new `StateKind` approch as a follow-up.

Reply:
Filed https://github.com/dotnet/roslyn/issues/65330

Reply:
> Let me open a separate tracking issue to try the new ```StateKind``` approch as a follow-up.

Sounds good